rootfs/api/tests/test_build.py

```diff
@@ -235,6 +236,144 @@ def test_build_default_containers(self, mock_requests):
+        url = "/v2/apps/{app_id}/builds".format(**locals())
+        body = {'image': 'autotest/example'}
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 201, response.data)
+
+        # verify worker is still there
+        url = "/v2/apps/{app_id}/pods/worker".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(len(response.data['results']), 0)
+
+        # verify web is still there
+        url = "/v2/apps/{app_id}/pods/web".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(len(response.data['results']), 0)
+
+        # look at the app structure
+        url = "/v2/apps/{app_id}".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(response.json()['structure'], {'web': 0, 'worker': 0})
+
+    @override_settings(PROCFILE_REMOVE_PROCS_ON_DEPLOY="")
+    def test_build_no_remove_process(self, mock_requests):
+        """
+        Specifically test PROCFILE_REMOVE_PROCS_ON_DEPLOY being turned off
+        and a user posting a new build without a Procfile that was previously
+        applid
+        """
+        # start with a new app
+        app_id = self.create_app()
+
+        # post a new build with procfile
+        url = "/v2/apps/{app_id}/builds".format(**locals())
+        body = {
+            'image': 'autotest/example',
+            'sha': 'a'*40,
+            'procfile': json.dumps({
+                'web': 'node server.js',
+                'worker': 'node worker.js'
+            })
+        }
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 201, response.data)
+
+        # scale worker
+        url = "/v2/apps/{app_id}/scale".format(**locals())
+        body = {'worker': 1}
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 204, response.data)
+
+        # verify worker
+        url = "/v2/apps/{app_id}/pods/worker".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(len(response.data['results']), 1)
+        container = response.data['results'][0]
+        self.assertEqual(container['type'], 'worker')
+        self.assertEqual(container['release'], 'v2')
+        # pod name is auto generated so use regex
+        self.assertRegex(container['name'], app_id + '-worker-[0-9]{8,10}-[a-z0-9]{5}')
+
+        # do another deploy for this time forget Procfile
+        url = "/v2/apps/{app_id}/builds".format(**locals())
+        body = {'image': 'autotest/example'}
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 201, response.data)
```

Comment:
@felipejfc No, there wouldn't be a "Before / After" view - We are not planning on adding client flags that only pertain to particular server side features. At least I am not convinced that is what we want to do.

The way I described would work for your particular use case was that if you submit a build with a Procfile and then submit another build without it then Deis would let that change through without removing any process types

However if you do the first build with a Procfile that contains `cmd`, `worker` and `redis` and then someone else submits a Procfile with `cmd` and `worker` then `redis` gets deleted, which is the normal Deis behaviour.

Basically that `server flag` for your scenario would be forgiving only when people forget the `Procfile` - Does that make sense?? - It's getting a little confusing since we have 2 similar but distinctly different outcomes being discussed in this thread ;-)

Just FYI the usage scenario @heynemann has is to fully *reject* if you go from Procfile to no Procfile.